core/src/main/java/io/grpc/ServerImpl.java

```diff
@@ -249,6 +257,7 @@ public void transportTerminated() {
     @Override
     public ServerStreamListener streamCreated(final ServerStream stream, final String methodName,
         final Metadata.Headers headers) {
+      final Future<?> timeoutFuture = scheduleTimeout(stream, headers);
```

Comment:
This timeout is not for the single call, but for the entire `ServerStream` life. You likely need to pass the `Future` to `ServerCallImpl` in `startCall()`. And then cancel it in `ServerStreamListenerImpl.closed()`. There may be some interesting threading involved.